### PR TITLE
Close #27023: Add capability to override telemetry URL

### LINKED ISSUE #27023: Add capability to override telemetry URL
### What is the user problem or growth opportunity you want to see solved?

Collecting telemetry from Fenix now is a bit involved to do locally. The steps are outlined [here](https://docs.google.com/document/d/1WqG9YnVnMJ3hk7vcym-kmxT2bsx92HqDagtl3nMG9wM/edit#heading=h.pazw2yi5ih1r). Allowing for a URL to be specified to send telemetry too, would help simplify this use case.

### How do you know that this problem exists today? Why is this important?

This isn't necessarily a problem but more of an enhancement to aid in testing/QA work.

### Who will benefit from it?

Softvision QA Team, Nimbus Engineering Team

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21589)

### PR DESCRIPTION
**Pull request by rocketsroger**
Tested with help from @jrbenny35.  I have confirmed that it works with a local [ping server](https://github.com/mozilla/experimenter/tree/main/app/tests/integration/nimbus/utils/ping_server).

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #27023

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt [hunk 1]
```diff
@@ -154,18 +156,38 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
...
-                ),
-            ),
-            uploadEnabled = telemetryEnabled,
-            buildInfo = GleanBuildInfo.buildInfo,
+        val customGleanServerUrl = PreferenceManager.getDefaultSharedPreferences(this).getString(
+            getPreferenceKey(R.string.pref_key_custom_glean_server_url),
+            "",
         )
 
+        if (customGleanServerUrl?.isNotBlank() == true) {
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:164]: Please do this this check a separate method too. There is a lot of in-lined logic in the FenixApplication class, we shouldn't add to it.

[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:164]: Sure, I'll point to the constant defined in the Glean service as default.  Thanks

**Comment by jonalmeida** [comment]
Roger and I spoke offline about this patch and we're going to update it to move some parts of the logic around.

<details>

<summary>WIP patch</summary>

```diff
diff --git a/app/src/main/java/org/mozilla/fenix/FenixApplication.kt b/app/src/main/java/org/mozilla/fenix/FenixApplication.kt
index 912205bf8..0df4c7fd0 100644
--- a/app/src/main/java/org/mozilla/fenix/FenixApplication.kt
+++ b/app/src/main/java/org/mozilla/fenix/FenixApplication.kt
@@ -5,6 +5,7 @@
 package org.mozilla.fenix
 
 import android.annotation.SuppressLint
+import android.content.Context
 import android.net.Uri
 import android.os.Build
 import android.os.Build.VERSION.SDK_INT
@@ -45,7 +46,6 @@ import mozilla.components.lib.crash.CrashReporter
 import mozilla.components.service.fxa.manager.SyncEnginesStorage
 import mozilla.components.service.glean.Glean
 import mozilla.components.service.glean.config.Configuration
-import mozilla.components.service.glean.config.Configuration.Companion.DEFAULT_TELEMETRY_ENDPOINT
 import mozilla.components.service.glean.net.ConceptFetchHttpUploader
 import mozilla.components.support.base.facts.register
 import mozilla.components.support.base.log.Log
@@ -76,6 +76,7 @@ import org.mozilla.fenix.components.appstate.AppAction
 import org.mozilla.fenix.components.metrics.MetricServiceType
 import org.mozilla.fenix.components.metrics.MozillaProductDetector
 import org.mozilla.fenix.components.toolbar.ToolbarPosition
+import org.mozilla.fenix.ext.components
 import org.mozilla.fenix.ext.containsQueryParameters
 import org.mozilla.fenix.ext.getPreferenceKey
 import org.mozilla.fenix.ext.isCustomEngine
@@ -157,15 +158,17 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
 
         logger.debug("Initializing Glean (uploadEnabled=$telemetryEnabled})")
 
+        val customEndpoint = getCustomGleanServerUrlIfAvailable(this)
+        val configuration = Configuration(
+            channel = BuildConfig.BUILD_TYPE,
+            httpClient = ConceptFetchHttpUploader(
+                lazy(LazyThreadSafetyMode.NONE) { components.core.client },
+            ),
+        )
+
         Glean.initialize(
             applicationContext = this,
-            configuration = Configuration(
-                serverEndpoint = getCustomGleanServerUrlIfAvailable() ?: DEFAULT_TELEMETRY_ENDPOINT,
-                channel = BuildConfig.BUILD_TYPE,
-                httpClient = ConceptFetchHttpUploader(
-                    lazy(LazyThreadSafetyMode.NONE) { components.core.client },
-                ),
-            ),
+            configuration = configuration.setCustomEndpointIfAvailable(customEndpoint),
             uploadEnabled = telemetryEnabled,
             buildInfo = GleanBuildInfo.buildInfo,
         )
@@ -867,17 +870,19 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
             components.useCases.wallpaperUseCases.initialize()
         }
     }
+}
 
-    private fun getCustomGleanServerUrlIfAvailable(): String? {
-        val customGleanServerUrl = PreferenceManager.getDefaultSharedPreferences(this).getString(
-            getPreferenceKey(R.string.pref_key_custom_glean_server_url),
-            "",
-        )
+private fun getCustomGleanServerUrlIfAvailable(context: Context): String? {
+    return PreferenceManager.getDefaultSharedPreferences(context).getString(
+        context.getPreferenceKey(R.string.pref_key_custom_glean_server_url),
+        null,
+    )
+}
 
-        return if (customGleanServerUrl.isNullOrBlank()) {
-            null
-        } else {
-            customGleanServerUrl
-        }
+fun Configuration.setCustomEndpointIfAvailable(serverEndpoint: String?): Configuration {
+    if (!serverEndpoint.isNullOrEmpty()) {
+        return copy(serverEndpoint = serverEndpoint)
     }
+
+    return this
 }
```

</details>

**Comment by rocketsroger** [comment]
> Roger and I spoke offline about this patch and we're going to update it to move some parts of the logic around.
> WIP patch

Thanks for the discussion and WIP patch.  I'll update.

**Review by jonalmeida — APPROVED** [review]
Thanks for pushing this forward! 🎉

**Pull request merged**